cmake/mirrors/gitcode.cmake

```diff
@@ -0,0 +1,53 @@
+# Tengine (Download via commit id)
+ocv_update(TENGINE_GITCODE_PKG_MD5 1b5908632b557275cd6e85b0c03f9690)
+# TBB (Download from release page)
+ocv_update(TBB_GITCODE_RELEASE "v2020.2")
+ocv_update(TBB_GITCODE_PKG_NAME "tbb-${TBB_GITCODE_RELEASE}")
+ocv_update(TBB_GITCODE_PKG_MD5 4eeafdf16a90cb66e39a31c8d6c6804e)
+# ADE (Download from release page)
+ocv_update(ADE_GITCODE_RELEASE "v0.1.1f")
+ocv_update(ADE_GITCODE_PKG_NAME "ade-${ADE_GITCODE_RELEASE}")
+ocv_update(ADE_GITCODE_PKG_MD5 c12909e0ccfa93138c820ba91ff37b3c)
+
+#
+# extract commit id and package name from original usercontent download link,
+# and replace the original download link with the one from gitcode
+#
+macro(ocv_download_url_gitcode_usercontent)
+    string(REPLACE "/" ";" DL_URL_split ${DL_URL})
+    list(GET DL_URL_split 5 __COMMIT_ID)
+    list(GET DL_URL_split 6 __PKG_NAME)
+    set(DL_URL "https://gitcode.net/opencv/opencv_3rdparty/-/raw/${__COMMIT_ID}/${__PKG_NAME}/")
+endmacro()
+#
+# extract repo owner and repo name from original usercontent download link,
+# and replace the original download link & md5 with the one from gitcode
+#
+macro(ocv_download_url_gitcode_archive_commit_id)
+    string(REPLACE "/" ";" DL_URL_split ${DL_URL})
+    list(GET DL_URL_split 3 __OWNER)
+    list(GET DL_URL_split 4 __REPO_NAME)
+    set(DL_URL "https://gitcode.net/${__OWNER}/${__REPO_NAME}/-/archive/")
+    set(DL_HASH "${${DL_ID}_GITCODE_PKG_MD5}")
```

Comment:
To avoid false-positive replacement it makes sense to compare "original" hash first.

Scenarios:
- PR updates 3rdparty component. This hooks file is left untouched. We should not keep downloading old resources.
- TBB dependency could be configured through CMake parameters for any version (see [ocv_update() calls](https://github.com/opencv/opencv/blob/4.5.5/3rdparty/tbb/CMakeLists.txt#L8-L10)). This hook can't support all of them.

Reply:
> To avoid false-positive replacement it makes sense to compare "original" hash first.

Is "original" hash the commit hash of the release? Commit hashes are not saved in CMake files downloading ADE & TBB. If we are using commit hash, I think we should also make ADE & TBB download using commit hash in the same way as Tengine does to keep CMake code unified and simple.

> PR updates 3rdparty component. This hooks file is left untouched. We should not keep downloading old resources.

As far as I know, mirrored repos in gitcode.net are automatically synced everyday. So hook files can be updated once PR updates 3rdparty components. Or do you mean we should fall back to original source if a newer version is used?

> TBB dependency could be configured through CMake parameters for any version.

Did you mean line 3 to line 6 ?
```cmake
# TBB (Download from release page)
ocv_update(TBB_GITCODE_RELEASE "v2020.2")
ocv_update(TBB_GITCODE_PKG_NAME "tbb-${TBB_GITCODE_RELEASE}")
ocv_update(TBB_GITCODE_PKG_MD5 4eeafdf16a90cb66e39a31c8d6c6804e)
```

Reply:
TBB release from scripts:
```
$ cmake ... -DWITH_TBB=ON -DBUILD_TBB=ON
...
-- TBB: Download: v2020.2.tar.gz
...
--   Parallel framework:            TBB (ver 2020.2 interface 11102)
```

Manual TBB version selection:
```
$ cmake ... -DWITH_TBB=ON -DBUILD_TBB=ON \
  -DOPENCV_TBB_RELEASE=v2020.3 -DOPENCV_TBB_RELEASE_MD5=ea8fa4332f4bad10a75a361cba025380
...
-- TBB: Download: v2020.3.tar.gz
...
--   Parallel framework:            TBB (ver 2020.3 interface 11103)
```

The second case should not be silently hijacked by mirror hooks and retrieve incorrect version.
Hooks may fail configuration or just don't change anything to cause failure on problem environment.

Reply:
In the latest commit, custom TBB is downloaded from github.com by default.

Reply:
It is not about TBB only. It is about updating of any dependencies (without changes of mirror hooks code).
Mirror hooks should NOT silently downgrade dependencies with "new outdated" hashsum.
We need to check the "original" hash before replacing hashsum and URL.
If we downloading something else then we should emit warning and do nothing (until we have updated hooks code).